draft-ietf-mls-protocol.md

```diff
@@ -4213,6 +4209,61 @@ Regardless of how the client obtains the tree, the client MUST verify that the
 root hash of the ratchet tree matches the `tree_hash` of the GroupContext before
 using the tree for MLS operations.
 
+## Validating Proposals
+
+A group member creating a commit and a group member processing a commit
+MUST verify that the list of committed proposals is valid using one of the following
+procedures, depending on whether the commit is external or not.
+
+For a regular, i.e. not external, commit the list is invalid if it contains any of the
+following:
+
+* An individual proposal that is invalid as specified in {{proposals}}.
+
+* An Update proposal generated by the committer.
+
+* A Remove proposal that removes the committer.
+
+* Multiple Update and/or Remove proposals that apply to the same leaf. If the committer
+  has received multiple such proposals they SHOULD prefer any Remove received, or
+  the most recent Update if there are no Removes.
+
+* Multiple Add proposals that contain KeyPackages that represent the same client
+  according to the application (for example, identical KeyPackages or KeyPackages
+  sharing the same Credential).
+
+* Multiple PreSharedKey proposals that reference the same PreSharedKeyID.
+
+* Multiple GroupContextExtensions proposals.
+
+* A ReInit proposal together with any other proposal. If the committer has
+  received other proposals during the epoch, they SHOULD prefer them over the
+  ReInit proposal, allowing the ReInit to be resent and applied in a subsequent
+  epoch.
+
+* An ExternalInit proposal.
+
+* A proposal with a non-default proposal type that is not supported by some
+  members of the group that will process the Commit (i.e., members being added
+  or removed by the Commit do not need to support the proposal type).
+
+An application may extend the above procedure by additional rules, for example,
+requiring application-level permissions to add members, or rules concerning
+non-default proposal types.
+
+For an external commit, the list is valid if:
```

Comment:
It's slightly confusing that the above is a negative check, and this is a positive check.  But that might be unavoidable.  I might rephrase as something like:

... the list is valid if it contains only the following proposals (not necessarily in this order):

* Exactly one ExternalInit proposal
* At most one Remove proposal...
* Zero or more PSK proposals...
* No other proposals

(The last bullet is a little redundant, but it bears repeating.)

Reply:
(And maybe we should require order here!)